services/spar/src/Spar/API.hs

```diff
@@ -322,23 +331,46 @@ assertNoScimOrNoIdP teamid = do
 validateNewIdP ::
   forall m.
   (HasCallStack, m ~ Spar) =>
+  WireIdPAPIVersion ->
   SAML.IdPMetadata ->
   TeamId ->
   Maybe SAML.IdPId ->
   m IdP
-validateNewIdP _idpMetadata teamId mReplaces = do
+validateNewIdP apiversion _idpMetadata teamId mReplaces = withDebugLog "validateNewIdP" (Just . show . (^. SAML.idpId)) $ do
   _idpId <- SAML.IdPId <$> SAML.createUUID
   oldIssuers :: [SAML.Issuer] <- case mReplaces of
     Nothing -> pure []
     Just replaces -> do
-      idp <- wrapMonadClient (Data.getIdPConfig replaces) >>= maybe (throwSpar SparIdPNotFound) pure
+      idp <- wrapMonadClient (Data.getIdPConfig replaces) >>= maybe (throwSpar (SparIdPNotFound (cs $ show mReplaces))) pure
       pure $ (idp ^. SAML.idpMetadata . SAML.edIssuer) : (idp ^. SAML.idpExtraInfo . wiOldIssuers)
   let requri = _idpMetadata ^. SAML.edRequestURI
-      _idpExtraInfo = WireIdP teamId oldIssuers Nothing
+      _idpExtraInfo = WireIdP teamId (Just apiversion) oldIssuers Nothing
   enforceHttps requri
-  wrapMonadClient (Data.getIdPIdByIssuer (_idpMetadata ^. SAML.edIssuer)) >>= \case
-    Nothing -> pure ()
-    Just _ -> throwSpar SparNewIdPAlreadyInUse
+  idp <- wrapMonadClient (Data.getIdPConfigByIssuer (_idpMetadata ^. SAML.edIssuer) teamId)
+  SAML.logger SAML.Debug $ show (apiversion, _idpMetadata, teamId, mReplaces)
+  SAML.logger SAML.Debug $ show (_idpId, oldIssuers, idp)
+
+  let handleIdPClash :: Either SAML.IdPId IdP -> m ()
+      handleIdPClash (Right idp') = case apiversion of
+        WireIdPAPIV1 -> do
+          throwSpar $ SparNewIdPAlreadyInUse "you can't create an IdP with api-version v1 if the issuer is already in use on the wire instance."
+        WireIdPAPIV2 -> do
+          when (fromMaybe defWireIdPAPIVersion (idp' ^. SAML.idpExtraInfo . wiApiVersion) == WireIdPAPIV1) $ do
```

Comment:
in fact, this is not redundant, but wrong.  instead of `idp'`, we should look at the version of the idp about to be replaced.  that could conceivable be v1, while we're replacing it with a v2, and that's not supported because i haven't thought it through.

Reply:
379c56c865ff36e3f5112882d5217bf26be17565